### PR TITLE
Add get_kernel_keys and eeprom_util

### PR DESCRIPTION
**Pull request by SIN3R6Y**
get_kernel_keys.py is a pretty simple tool, not too much to explain.

eeprom_util.py could be expanded upon in the future by allowing it to change the data in the eeprom. Maybe plug into the kernel to live change it on a system if possible? But for now it just prints some basic data and prints it to shell. It also verifies that the decryption was valid. See comments for more details.

### CODE REVIEW

File: python-scripts/eeprom-util.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    hashData.append(byte)
+  for byte in decData:
+    hashData.append(byte)
+
+  # hash decrypted EEPROM data
+  eepHash = hmac.new(eepKey, hashData, sha1).digest()
+
+  # make sure hashes match
+  if eepHash == eeprom.hash:
+    print ('\n\033[1mThe decrypted and encrypted EEPROM hashes match!\033[0m')
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: does `\033[1m` work on all platforms / what does it do? Maybe there is a python module (maybe even included with python already) which could handle this? If not, please put these into a meaningful variable name and concatenate it or something.

File: python-scripts/eeprom-util.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  eepKeyHash = hmac.new(eepKey, eeprom.hash, sha1).digest()
+
+  # initialize EepKeyObj method object
+  eepKeyObj = EepKeyObj()
+  eepKeyPrepCrypt(eepKeyHash, eepKeyObj)
+  
+  # decrypt confounder 
+  decConf = eepKeyCrypt(eeprom.confounder, eepKeyObj)
+  
+  # assemble encrypted HDD key and game region code into a single bytearray
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: Why does this have to be done? does the original kernel also do this? Would it work to just decrypt the EEPROM image *before* parsing it?

File: python-scripts/eeprom-util.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    self.unknown5 = bytes(eepFile.read(64))       # TBD
+
+# Because of the confounder we have to split the RC4 decryption / encryption proccess
+# into two functions (eepKeyPrepCrypt / eepKeyCrypt). We also need a way to share state
+# between the the two functions (EepKeyObj). In a nutshell we fill and EepKeyObj method
+# object with the initial state by calling eepKeyPrepCrypt. Then we call eepKeyCrypt
+# on the confounder data. eepKeyCrypt will push state data back into the EepKeyObj method 
+# object (i / j) which is necessary when we call eepKeyCrypt on the next chunk of data
+# from the Eeprom method object (eeprom.hddKey + eeprom.gameReg). This is also why we
+# don't use any standard RC4 crypto libaries here.
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: I'm not sure if I understand the issue. A standard RC4 implementation should also support this: https://wiki.openssl.org/index.php/Manual:Rc4(3)

In case it doesn't, please still mark / name your functions as being RC4ish and link to a source / explanation of the algorithm (Wikipedia might be fine) and pointing out the differences / details in usage.

File: python-scripts/eeprom-util.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  print ('Encrypted Game Region Code:\t{}'.format(''.join(format(x, '02x') for x in eeprom.gameReg)))
+
+  print ('\n\033[1mIdentity EEPROM Data\033[0m')
+  print ('Identity Checksum:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.identChk)))
+  print ('Serial Number:\t\t\t{}'.format(eeprom.serial.decode('ascii')))
+  print ('MAC Address:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.mac)))
+  print ('Online Key:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.onlineKey)))
+
+  # print NTSC / PAL instead of their respective codes
+  if ''.join(format(x, '02x') for x in eeprom.videoStd) == '00014000':
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: why the string compare? I'd advise converting to int or comparing bytes objects
However, you are better at python probably, sooo.. why? :P

File: python-scripts/eeprom-util.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    print ('Video Standard:\t\t\tNTSC')  
+  elif ''.join(format(x, '02x') for x in eeprom.videoStd) == '00038000':
+    print ('Video Standard:\t\t\tPAL')
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: A tool like this has the purpose of being able to avoid a hex-editor. If the user is unsure about the actual contents they might still need to hex-edit in certain situations. That should be avoided.

With the check testing for very specific values (which look like bitmasks or something), we should always print the actual value; and then *additionally* tell the user what we understood the value as.

File: python-scripts/get_kernel_keys.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                  help='view all keys found in kernel memory space')
+(options, args) = parser.parse_args()
+
+eepKey = bytes(read(ke.XboxEEPROMKey(), 16, False))
+hddKey = bytes(read(ke.XboxHDKey(), 16, False))
+sigKey = bytes(read(ke.XboxSignatureKey(), 16, False))
+lanKey = bytes(read(ke.XboxLANKey(), 16, False))
+
+if int.from_bytes(eepKey, 'big', signed=False) == 0:
+  print ("\n\033[1mWARNING: Your EEPROM Key has been erased! Is your xbox running a retail BIOS?")
```
Comment (JayFoxRox) [inline python-scripts/get_kernel_keys.py]: User visible spelling of `Xbox` should be `Xbox` (capital `X`, lowercase `box`; no other symbols)

File: python-scripts/eeprom-util.py (hunk 3 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: https://en.wikipedia.org/wiki/RC4 - This was written based on that

You have two RC4 stages, the KSA and PRGA. The KSA is initializing a permutation of the key, The PRGA actually does the encryption / decryption. Consider the following based on a standard RC4 algorithm.

Normally It goes like this.

```
def rc4 (key, data):
  keyPerm = KSA(key) # generates a permutation using the key
  decData = PGRA(keyPerm, data) # decrypts data using the key permutation 
  return decData
```

So if we try to encrypt an eeprom with this function this is what happens.

```
confounder = rc4 (eepromKey, encryptedConfouder)

def rc4 (key, data):
  keyPerm = KSA(key)
  decData = PGRA(keyPerm, data)
  return decData
```

This part works, we get our confounder, but doing it again will not work for the hdd key / game reg

```
hddKeyGameReg = rc4 (eepromKey, encryptedBytes)

def rc4 (key, data):
  #after the last function ended we lost the key permutation, no issues we'll make it again
  #except during the PGRA stage we stored I and J from decrypting the confounder in the
  # permutation. We need I and J from the confounder PGRA stage to decrypt further .
  # So keyPerm needs to be a seperate entitiy. Also KSA need to only be called once or
  # it wipe data from the previous PGRA stage as well.
  keyPerm = KSA(key)

  # By default this function set I and J to zero at start, we have to change that as it needs to 
  # pull those values from keyPerm. If you don't do this it wont work.
  decData = PGRA(keyPerm, data)
  return decData
```

So keyPerm, KSA, and PGRA all must be changed from their default behaviour in order to account for the confounder.

keyPerm data has to persist between function calls, so it must sit outside of the functions.
KSA must only be once, when we first decrypt the confounder
PGRA must pull starting values from keyPerm and export the finishing values back to keyPerm

It's hard to follow i know, i beat my head against a wall four hours figuring it out, but that's why the standard RC4 functions don't work, because they don't consider that the confounder data has to stick around through the rest of the process.

File: python-scripts/eeprom-util.py (hunk 4 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: No reason, i was formatting everything as a string there so i probably didn't even think about it

File: python-scripts/eeprom-util.py (hunk 1 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: It just makes the text bold, should work on any system supporting utf-8

File: python-scripts/eeprom-util.py (hunk 2 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: The process is decrypt the confounder which sets up eepKeyObj to be able to decrypt the HDD key / game region code. Those two pieces of data were originally encrypted as one 20 byte chunk, so we have to decrypt it that way.

File: python-scripts/eeprom-util.py (hunk 1 above)
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: It's not unicode. If it was unicode it would change every character.

[This is probably a VT100 escape code](http://www.climagic.org/mirrors/VT100_Escape_Codes.html). I've worked on other projects which had hundreds of lines of code to handle different terminals. Windows in particular can be troublesome (I think they use a custom API instead of VT100).

File: python-scripts/eeprom-util.py (hunk 2 above)
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: yes, but isn't it easier to decrypt or hash these fields *before* unpacking them with read(4) etc.?

File: python-scripts/eeprom-util.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+  # decrypt HDD key and game region code
+  decData = eepKeyCrypt(eepData, eepKeyObj)
+
+  # assemble decrypted EEPROM data into a single bytearray
+  hashData = bytearray()
+  for byte in decConf:
+    hashData.append(byte)
+  for byte in decData:
+    hashData.append(byte)
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: Wouldn't `hashdata = decConf + decData` work? other instances too.

File: python-scripts/eeprom-util.py (hunk 2 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: I believe read() is a one way street. As the file buffer automatically moves to the next byte for ever read(n) call. We would have to close the file, re open it, skip the first n bytes to the data we want, then read again. Unless my understand of pythons read() is wrong when in byte mode

File: python-scripts/eeprom-util.py (hunk 7 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: Maybe, ill give that a try. Im used to python 2 where your only option is a bytearray and iterating through it.

File: python-scripts/eeprom-util.py (hunk 3 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: Comments documents each functions have been added

File: python-scripts/eeprom-util.py (hunk 4 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: changed to int compare, python byte notation is ugly look. With a few comments it's easy to know what's going on. I probably did them as strings originally because you could easily see what it was doing.

File: python-scripts/eeprom-util.py (hunk 5 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: Added hex values in parentheses

File: python-scripts/eeprom-util.py (hunk 2 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: Changed to load EEPROM dump into variable, then closing the file as we no longer need it. Using standard byte operators to assign the data to the EEPROM class.

File: python-scripts/eeprom-util.py (hunk 7 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: Yes this works, fixed

File: python-scripts/eeprom-util.py (hunk 1 above)
Comment (SIN3R6Y) [inline python-scripts/eeprom-util.py]: I removed them, it looked nice but there was no easy way to make it 100% cross platform

File: python-scripts/get_kernel_keys.py (hunk 6 above)
Comment (SIN3R6Y) [inline python-scripts/get_kernel_keys.py]: Fixed

File: python-scripts/eeprom-util.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    self.gameParCont = eepDump[156:160]     # game parental contol settings
+    self.parContPass = eepDump[160:164]     # parental control password
+    self.dvdParCont = eepDump[164:168]      # DVD parental control settings
+    self.ip = eepDump[168:172]              # IP address
+    self.dns = eepDump[172:176]             # DNS server
+    self.gateway = eepDump[176:180]         # default gateway
+    self.subnet = eepDump[180:184]          # subnet
+    self.unknown4 = eepDump[184:188]        # TBD
+    self.dvdReg = eepDump[188:192]          # DVD region code
+    self.unknown5 = eepDump[192:256]       # TBD
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: nit: please align comments properly

File: python-scripts/eeprom-util.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  print ('\nIDENTITY EEPROM DATA')
+  print ('Identity Checksum:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.identChk)))
+  print ('Serial Number:\t\t\t{}'.format(eeprom.serial.decode('ascii')))
+  print ('MAC Address:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.mac)))
+  print ('Online Key:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.onlineKey)))
+
+  # print NTSC / PAL instead of their respective codes
+  if int.from_bytes(eeprom.videoStd, 'big', signed=False) == 81920: # 81920 == 00014000 
+    print ('Video Standard:\t\t\tNTSC ({})'.format(''.join(format(x, '02x') for x in eeprom.videoStd)))  
+  elif int.from_bytes(eeprom.videoStd, 'big', signed=False) == 229376: # 229376 == 00038000
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: nit: please use hex constants

File: python-scripts/eeprom-util.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    print ('Video Standard:\t\t\tPAL ({})'.format(''.join(format(x, '02x') for x in eeprom.videoStd)))
+  else:
+    print ('Video Standard:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.videoStd)))
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: nit: Consider splitting video mode detection and printing the video mode constant

```python
videoStd =  int.from_bytes(eeprom.videoStd, 'little', signed=False)
if videoStd == 0x1337: detectedVideoStd = "PAL"
elif videoStd == 0x1338: detectedVideoStd = "NTSC"
else: detectedVideoStd = "Unknown"
print (('Video Standard:\t\t\t0x%08X' % videoStd) + " (" detectedVideoStd + ")")
```

(Something like that; untested code as example only)

File: python-scripts/eeprom-util.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+      print ('Decrypted Confounder:\t\t{}'.format(''.join(format(x, '02x') for x in decConf)))
+      print ('Decrypted HDD Key:\t\t{}'.format(''.join(format(x, '02x') for x in decHddKey)))
+      print ('Decrypted Game Region Code:\t{}'.format(''.join(format(x, '02x') for x in decGameReg)))
+
+    # hashes don't match
+    else:
+      print ('\nWARNING: The decrypted and encrypted EEPROM hashes do not match!')
+      print ('Either your EEPROM file or EEPROM key is corrupt / invalid!\n')
+      print ('Original EEPROM Hash:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.hash)))
+      print ('Invalid EEPROM Hash:\t\t{}'.format(''.join(format(x, '02x') for x in eepHash)))
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: nit: A hash mismatch should not prevent us from outputting additional data (such as the keys) in my opinion.
As long as it's obvious that there was a mismatch (and which data it applies to / which data might be broken or wrong) we should give as much information as possible.

File: python-scripts/get_kernel_keys.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+parser.add_argument('-s', action='store_true', dest='sec',
+                  help='view security keys found in kernel memory space')
+parser.add_argument('-o', action='store_true', dest='other',
+                  help='view other keys found in kernel memory space')
+parser.add_argument('-x', action='store_true', dest='exec',
+                  help='view executable keys found in kernel memory space')
+args = parser.parse_args()
+
+eepKey = read(ke.XboxEEPROMKey(), 16, False)
+hddKey = read(ke.XboxHDKey(), 16, False)
```
Comment (JayFoxRox) [inline python-scripts/get_kernel_keys.py]: from xbox-dump: ` uint8_t* XboxCERTKey = &XboxHDKey[-16]; // XboxCERTKey is just infront of XboxHDKey`
Or translated to python: `certKey = read(ke.XboxHDKey() - 16, 16, False) # XboxCERTKey is just infront of XboxHDKey`

This key should be present, and be at that location for almost all kernels. We should also output it.
If possible, it should be warned that the CERT key might not be correct [as it's purely a guess where it is located in memory].

iirc it's one of the keys being passed to the kernel at startup, so it's actually an important key for emulation.

File: python-scripts/get_kernel_keys.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+import argparse
+import textwrap
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-d', action='store_true', dest='dump',
+                  help='dump security keys to a file (combine with -s)')
+parser.add_argument('-s', action='store_true', dest='sec',
+                  help='view security keys found in kernel memory space')
+parser.add_argument('-o', action='store_true', dest='other',
+                  help='view other keys found in kernel memory space')
```
Comment (JayFoxRox) [inline python-scripts/get_kernel_keys.py]: Please seperate kernel constant and game (title-app / running XBE) specific keys [to the best of your knowledge] or don't seperate them at all.
The current way of seperating / grouping keys is just confusing.

**Comment by JayFoxRox** [comment]
Note: After merge, we should have an issue about extending the eeprom tool to download the eeprom image from an Xbox via network (or a seperate python script for that). Alternatively, we could move the eeprom parsing into the xbox/ module in the future.

The current directory structure of this repo is that all tools in the "python-scripts" folder interface with a real Xbox. All other folders contain small helper tools to operate on dumps or be ran on Xbox stand-alone.
The tool is currently too stand-alone to belong in the python-scripts folder, but not generic enough for the xbox module.

@SIN3R6Y please move the eeprom-util into a seperate folder called "eeprom-parser" (or something along those lines). The key script is in the correct folder still.

File: python-scripts/eeprom-util.py (hunk 9 above)
Comment (SIN3R6Y) [reply python-scripts/eeprom-util.py]: No, little endian will reverse the byte order, this is stored in memory already in the correct order

File: python-scripts/eeprom-util.py (hunk 11 above)
Comment (SIN3R6Y) [reply python-scripts/eeprom-util.py]: If the has matches, the key was wrong, meaning the whole decryption was wrong. It would be useless information to know all the wrong contents?

File: python-scripts/eeprom-util.py (hunk 9 above)
Comment (SIN3R6Y) [reply python-scripts/eeprom-util.py]: To use hex contents we'd have to convert to string as hex() expects a string in python. We could use byte notation, but that will look really ugly. It used to be string, but i changed it due to a previous complaint.

Comment (JayFoxRox) [reply python-scripts/eeprom-util.py]:
```
[fox@x230 ~]$ python3
Python 3.6.2 (default, Jul 20 2017, 03:52:27) 
[GCC 7.1.1 20170630] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> print(0x100)
256
```

**Comment by JayFoxRox** [comment]
@SIN3R6Y Reminder about this. I'd love to have this but currently don't have the time to do a final cleanup

File: python-scripts/eeprom-util.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+  return eepData
+
+parser = optparse.OptionParser()
+parser.add_option('-f', action='store', dest='eepFileLoc')
+parser.add_option('-k', action='store', dest='keyFileLoc')
+(options, args) = parser.parse_args()
+
+if options.eepFileLoc:
+  eeprom = Eeprom(options.eepFileLoc)
+  print ('\033[1mEncrypted EEPROM Data\033[0m')
```
Comment (JayFoxRox) [inline python-scripts/eeprom-util.py]: Please remove these terminal control codes, they might mess up on some platforms.

**Comment by SIN3R6Y** [comment]
@JayFoxRox going to be a few days, but ill get to it.

**Comment by JayFoxRox** [comment]
In retrospect I probably should have merged an earlier version of this almost straight away.
Having something to iterate on is better than having nothing at all :(

I'll close this for now. Please re-open after re-base (or possibly submit PR to xboxpy or something).

**Comment by SIN3R6Y** [comment]
Yea i've been thinking about this lately, i'll get with you sometime before the end of the year and try to clean this up and make a new PR
